### PR TITLE
Feature/redbox 411 chunk refactor in api

### PR DESCRIPTION
**Pull request by jamesrichards4**
## Context

There is an assumption throughout the code that there is a file and a set of chunks. This will not be true long term. This refactor removes the strong link between files and chunks and tries to use Langchain retrievers to interact with chunks where possible.

## Changes proposed in this pull request

Use retrievers for all chunk interactions (see below for why this is a lie)
Avoid use of Chunk class in 'AI' code, chains and runnables handle Documents which lines up with Langchain stuff
Added additional mapping functions for document management vs chunks/source documents
Explicitly setting file processing status on the file in Elasticsearch. This means we can update the file status without the API interpreting it from chunks. This will help if/when we have multiple sets of chunks for additional processing beyond a single index of chunks. We now return processing and complete statuses to Django.

## Guidance to review

The StorageHandler class is still needed for file interactions, this is clean and makes sense (particularly because we may want to make more use of this class). It is also needed to manage returning chunks and statuses in the API. File status provides a list of chunk statuses, this will need revisiting and if removed could allow further clean up.

The chunk clustering code is still present in case we want to restore that functionality (it is now only a code skeleton as it hasn't been mapped to the new Document setup). We could remove all of redbox.parsing if we decide not to do this clustering. Dropping this would be my preference!

There is a lot of messy document mapping happening. Tracking this through it is caused by the rechunking happening in the summarisation chains. If we create summarisation chunks at ingest time (second worker) we could avoid all of this and then only the retrievers need to understand the document structure

The new retrievers need tests, I'll add these on this PR but I wanted to raise it to get thoughts about the structure

## Things to check

There's a lot messing around in the retrievers to handle old style documents. This should allow us to maintain compatibility with documents ingested by the old worker (chunk style). Worth noting and also confirming that this works in dev!

Can Django receive the new file status of 'processing'? I imagine not?

- [ ] I have added any new ENV vars in all deployed environments
- [ ] I have tested any code added or changed
- [ ] I have run integration tests

### CODE REVIEW

File: redbox/models/file.py [hunk 1]
```diff
@@ -13,8 +13,8 @@
 
 
 class ProcessingStatusEnum(str, Enum):
-    chunking = "chunking"
-    embedding = "embedding"
+    processing = "processing"
+    failed = "failed"
```
[the remark this answers is not in the document]
Comment (jamesrichards4) [reply redbox/models/file.py:17]: I suspected so, there's an enum in the Django app but I wanted to check with those who actually understand that app!

[the remark this answers is not in the document]
Comment (brunns) [reply redbox/models/file.py:17]: We'd need to do this with 3 migrations - one to add the new statuses, one to update files using the old ones to the new ones, and one to remove the old statuses. Plus we'd need to fix the document and chat views.

Shall I jump on to this branch and make those changes?

[the remark this answers is not in the document]
Comment (gecBurton) [reply redbox/models/file.py:17]: for the time being could we just make this a string in django @brunns ? This significant refactor isnt over yet an:
1. there could be more changes
2. id like to decouple the front and back ends

File: core_api/src/format.py [hunk 2]
```diff
@@ -1,49 +1,41 @@
...
+def format_documents(documents: list[Document]) -> str:
     formatted: list[str] = []
-
-    for chunk in chunks:
-        doc_xml = f"<Doc{chunk.parent_file_uuid}>\n {chunk.text} \n</Doc{chunk.parent_file_uuid}>"
+    for d in documents:
+        parent_file_uuid = d.metadata.get("parent_file_uuid")  # New Style Ingest
+        if not parent_file_uuid:
+            parent_file_uuid = d.metadata.get("parent_doc_uuid")  # Old Style Ingest
+        doc_xml = f"<Doc{parent_file_uuid}>\n {d.page_content} \n</Doc{parent_file_uuid}>"
```
Comment (gecBurton) [inline core_api/src/format.py:13]: is this XML format required by langchain?

File: core_api/src/format.py [hunk 3]
```diff
@@ -1,49 +1,41 @@
...
 
-def reduce_chunks_by_tokens(chunks: list[Chunk] | None, chunk: Chunk, max_tokens: int) -> list[Chunk]:
+def reduce_chunks_by_tokens(chunks: list[Document] | None, chunk: Document, max_tokens: int) -> list[Document]:
     if not chunks:
         return [chunk]
 
-    last_chunk = chunks[-1]
+    # Backwards compatible with worker which didn't store token_count
+    # Everything is optional None so we have to check everything
+    # This will all be rolled up into a SummarisationChunkRetriever or similar in future work and this ugliness
```
Comment (gecBurton) [inline core_api/src/format.py:25]: thx for explanation 👍

File: core_api/src/routes/file.py [hunk 4]
```diff
@@ -251,16 +251,23 @@ def get_file_status(file_uuid: UUID, user_uuid: Annotated[UUID, Depends(get_user
...
-    except ValueError:
-        return file_not_found_response(file_uuid=file_uuid)
-
-    return status
+    if file.ingest_status is not None:
+        return FileStatus(
+            file_uuid=file_uuid,
+            # We need to break the link between file status and a specific set of chunks
+            # to enable future work with many chunks or other indices etc
+            chunk_statuses=[],
```
Comment (gecBurton) [inline core_api/src/routes/file.py:266]: im wondering if we shouldnt just get rid of chunk status all together?

File: core_api/tests/conftest.py [hunk 5]
```diff
@@ -9,11 +12,22 @@
...
 
 from core_api.src.app import app as application
 from redbox.model_db import MODEL_PATH
 from redbox.models import Chunk, File, Settings
+from redbox.models.file import Metadata
 from redbox.storage import ElasticsearchStorageHandler
+from core_api.src.retriever import get_all_chunks_query, ParameterisedElasticsearchRetriever
+from core_api.src.retriever.parameterised import parameterised_document_mapper
+
+ALL_CHUNKS_RETRIEVER_RESOURCES = json.load(open("core_api/tests/resources/retriever/all_chunks.json"))
```
Comment (gecBurton) [inline core_api/tests/conftest.py:29]: pls can we just encode these as python? i dont think we are gaining anything by having them in json and it is conceivable (if unlikely) that the IDE might be able to help with any errors in the params.

File: core_api/tests/retriever/test_parameterised.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+from langchain_elasticsearch.retrievers import ElasticsearchRetriever
+from langchain_core.documents.base import Document
+
+from core_api.src.retriever import ParameterisedElasticsearchRetriever
+from redbox.models.file import Chunk, File
+
+
+def test_parameterised_retriever(parameterised_retriever: ParameterisedElasticsearchRetriever, stored_file_parameterised: List[Document]):
+    for k in [1, 2]:
```
Comment (gecBurton) [inline core_api/tests/retriever/test_parameterised.py:13]: could be a test parameter?

**Review by gecBurton — APPROVED** [review]
TLDR : I like the approach

I've added some nitpicking but this can be ignored, better to press ahead to the next stage and tidy up later. I see there are some empty files added too

v keen to have @brunns thoughts on this top spot breaking changes to the front end

File: redbox/models/file.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (brunns) [reply redbox/models/file.py:17]: We could do, bit we base a bit of behavior on this value; we don't just display it. Values for this that Django doesn't know about will probably cause problems, and I think the cause will be more obvious if we're using an enum.

We can certainly decouple `core-api` and `django-app` a bit, though, by agreeing on the message. Only 3 values are really useful here - prosessing, ready, and it's-gone-pete-tong. If we align on that, what `core-api` and `django-app` do internally can be their own business.

[the remark this answers is not in the document]
Comment (jamesrichards4) [reply redbox/models/file.py:17]: the behaviour is presumably when to enable chat on it? it has to be 'complete' for that I guess? It's only the processing status that's new so can't we leave the various statuses in place? Aren't parsing, chunking, embedding and now processing all the same to the frontend?

[the remark this answers is not in the document]
Comment (brunns) [reply redbox/models/file.py:17]: Basically, yes. So if we could not send core's internal idea of the status, but just send one of three values, that would isolate the 2 services. 

If we continue to expose core's internal status, we'll keep needing to change core & django at the same time. Let's say you added, I don't know, __munching__ as a status - we'd need a change on the django side to know what to do with it.

File: core_api/src/retriever/parameterised.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
+from functools import partial
+from typing import Any, TypedDict, Dict
```
[the remark this answers is not in the document]
Comment (jamesrichards4) [reply core_api/src/retriever/parameterised.py]: oh that is good to know!

File: redbox/transform.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+        file_uuid=(
+            d.metadata["_source"]["metadata"]["parent_file_uuid"]
+            if "parent_file_uuid" in d.metadata["_source"].get("metadata", {})
+            else d.metadata["_source"].get("metadata", {}).get("parent_doc_uuid")
+        ),
+        page_numbers=map_page_numbers(d.metadata["_source"]["metadata"].get("page_number")),
+    )
+
+
+# This should be unnecessary and indicates we're not chunking correctly
```
[the remark this answers is not in the document]
Comment (jamesrichards4) [reply redbox/transform.py:27]: I don't love this but as the new document metadata is copied from the first doc passed in 'a' all values are implicitly the same. Values which want a combined value are then explicitly mapped. This creates a weird 'no mapping' for some values in that function which is harder to parse then completely explicit

File: worker/src/app.py [hunk 9]
```diff
@@ -69,12 +72,29 @@ def wrapped(file: File):
...
-    ).invoke(file)
+    file.ingest_status = ProcessingStatusEnum.processing
+    storage_handler.update_item(file)
+
+    try:
+        new_ids = (
+            document_loader(s3_client=s3_client, env=env)
+            | RunnableLambda(list)
+            | RunnableLambda(vectorstore.add_documents)
+        ).invoke(file)
```
[the remark this answers is not in the document]
Comment (jamesrichards4) [reply worker/src/app.py:91]: yes you're right and vectorstore supports async too (aadd_documents) so we could probably improve this going that way. I'll bump that to another ticket though unless you have strong feelings?

NOTE: in general I'm not sure where to record stuff like this, product seem to manage our tickets and this is internal so not really on them. Do we just slap stuff on the backlog and make sure it gets pulled??

File: worker/src/app.py [hunk 10]
```diff
@@ -69,12 +72,29 @@ def wrapped(file: File):
...
+        file.ingest_status = ProcessingStatusEnum.complete
+    except Exception:
```
[the remark this answers is not in the document]
Comment (jamesrichards4) [reply worker/src/app.py:93]: I definitely need to get better with lint/format. I'd be in favour of checks in pre-commits to catch that but happy to just try harder too!

File: core_api/src/retriever/__init__.py [hunk 11]
```diff
@@ -0,0 +1,4 @@
+from .all_chunks import get_all_chunks_query
+from .parameterised import ParameterisedElasticsearchRetriever
```
[the remark this answers is not in the document]
Comment (jamesrichards4) [reply core_api/src/retriever/__init__.py:2]: I don't like the layout I've ended up with for sure and your suggestion is much cleaner. I'd suggest that we thing about this when we add new ones and think more about the summarisation/large chunk retriever as there might be significant work here again?

File: core_api/src/retriever/parameterised.py [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+# This is used in tests to avoid an issue when using with_config
+# It is a simple doc mapper, equivalent to setting content_field to 'text'
+# We don't get it in the app so there's something weird going on with tests
+def parameterised_document_mapper(hit: Dict) -> Document:
+    return Document(
+        page_content=hit['_source']['text'],
+        metadata=hit['_source'].get('metadata', {})
+    )
+
+class ParameterisedElasticsearchRetriever(ElasticsearchRetriever):
```
[the remark this answers is not in the document]
Comment (jamesrichards4) [reply core_api/src/retriever/parameterised.py:82]: I agree that we should keep this. I would see the advantage in two main parts, we can start using the high level 'combination' retrievers easily now [citation needed] (ensemble, contextualcompression) and we can start adding in others where appropriate and see what happens. This low level one lets us flex our Elastic knowledge though and we'll likely always want that kind of control I think

File: core_api/src/runnables.py [hunk 13]
```diff
@@ -180,11 +179,30 @@ def make_condense_rag_runnable(
...
             "sources": retriever,
         }
         | {
             "response": prompt | llm | StrOutputParser(),
             "sources": itemgetter("sources"),
         }
     )
+
+
+def resize_documents(max_tokens: int | None = None) -> list[Document]:
```
[the remark this answers is not in the document]
Comment (jamesrichards4) [reply core_api/src/runnables.py:192]: we're 100% reinventing the wheel here and I want to think about a new ingest pipeline which creates these larger chunks. I think langchain can help here with the ParentDocumentRetriever as you say so that'd be good to look at. We can remove so much faffing around with docs if we get that working

File: redbox/tests/test_transform.py [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+                page_content="some random text2",
+                metadata={
+                    "_source": {
+                        "metadata": {
+                            "parent_file_uuid": "asdf",
+                            "creator_user_uuid": "xabcd",
+                            "index": 12,
+                            "page_number": 3,
+                            "languages": [],
+                            # "link_texts": [],
```
[the remark this answers is not in the document]
Comment (jamesrichards4) [reply redbox/tests/test_transform.py:151]: it's also kind of ugly but I think we can add lightweight wrappers as needed here (transform functions etc)

**Review by gecBurton — APPROVED** [review]
